### PR TITLE
SSH integration

### PR DESCRIPTION
**Pull request by andrewbuss**
The first commit adds an API endpoint to Red October which decrypts an SSH key and uses it to sign a message. 

The later commit uses this endpoint to implement a remote signer which is used in a minimal SSH agent implementation. The goal is to be able to use an unmodified SSH client. When the SSH client tries to authenticate to a remote server, the server asks the client to sign a message with its key to set up the session. The client here forwards the message to our custom agent, which in turn passes it to the RO server, which signs the message if all necessary delegations are available. The signature is passed back along to the server through the agent and SSH client. 

The agent needs to know which public key corresponds to the encrypted private key, so there is a -pubkey command line flag which accepts the path to the appropriate pubkey file. 

The third commit adds an optional Usages field to the EncryptedData struct. A user can specify on encryption that an encrypted file may only be used as an SSH signer, and may not be directly decrypted to return a private key. If Usages is missing or empty, decryption is permitted.

Example:

`ro -in id_rsa.encrypted -server redoctob.er:8080 -pubkey id_rsa.pub ssh root@server`

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -251,6 +251,22 @@ func (c *RemoteServer) Decrypt(req core.DecryptRequest) (*core.ResponseData, err
 
 }
 
+// Decrypt issues an decrypt request to the remote server
```
Comment (grittygrease) [inline client/client.go]: Should be DecryptSign

File: cmd/ro/roagent/roagent.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
+package roagent
```
Comment (grittygrease) [inline cmd/ro/roagent/roagent.go]: Package comments please.

File: cmd/ro/roagent/roagent.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	respSignature := ssh.Signature{Format:respMsg.SignatureFormat, Blob:respMsg.Signature}
+	return &respSignature, nil
+}
+
+type ROAgent struct {
+	signer ROSigner
+}
+
+func NewROAgent(server *client.RemoteServer, pubKey ssh.PublicKey, encryptedPrivKey []byte, user, pswd string) agent.Agent {
```
Comment (grittygrease) [inline cmd/ro/roagent/roagent.go]: Exported functions should have comments.

File: core/core_test.go [hunk 4]
```diff
@@ -761,7 +807,7 @@ func TestOwners(t *testing.T) {
 	delegateJson := []byte("{\"Name\":\"Alice\",\"Password\":\"Hello\",\"Time\":\"0s\",\"Uses\":0}")
 	delegateJson2 := []byte("{\"Name\":\"Bob\",\"Password\":\"Hello\",\"Time\":\"0s\",\"Uses\":0}")
 	delegateJson3 := []byte("{\"Name\":\"Carol\",\"Password\":\"Hello\",\"Time\":\"0s\",\"Uses\":0}")
-	encryptJson := []byte("{\"Name\":\"Carol\",\"Password\":\"Hello\",\"Minumum\":2,\"Owners\":[\"Alice\",\"Bob\",\"Carol\"],\"Data\":\"SGVsbG8gSmVsbG8=\"}")
+	encryptJson := []byte("{\"Name\":\"Carol\",\"Password\":\"Hello\",\"Minimum\":2,\"Owners\":[\"Alice\",\"Bob\",\"Carol\"],\"Data\":\"SGVsbG8gSmVsbG8=\"}")
```
Comment (grittygrease) [inline core/core_test.go]: Good catch. This may already have been fixed in a merged PR, so please rebase.

File: cryptor/cryptor.go [hunk 5]
```diff
@@ -116,6 +117,7 @@ type EncryptedData struct {
 	IV        []byte                      `json:",omitempty"`
 	Data      []byte
 	Signature []byte
+	Usages    []string `json:",omitempty"`
```
Comment (grittygrease) [inline cryptor/cryptor.go]: This looks like a gofmt issue.

File: cryptor/cryptor.go [hunk 6]
```diff
@@ -253,27 +255,27 @@ func (encrypted *EncryptedData) wrapKey(records *passvault.Records, clearKey []b
...
-		}
-
-		innerCrypt.Encrypt(keyBytes, clearKey)
-		outerCrypt.Encrypt(keyBytes, keyBytes)
-
 		return
 	}
 
 	if len(access.Names) > 0 {
+		if len(access.Names) < access.Minimum {
```
Comment (grittygrease) [inline cryptor/cryptor.go]: Lots of this stuff should go away with a rebase.

**Comment by grittygrease** [comment]
This could use a section in the README about how to set up and an example of how to use it.

File: client/client.go (hunk 1 above)
Comment (andrewbuss) [inline client/client.go]: Thanks - fixed in next push. I'm wondering if there's a better name for this endpoint - decrypt-sign could be confusing - but I couldn't come up with anything.

File: cryptor/cryptor.go (hunk 5 above)
Comment (andrewbuss) [inline cryptor/cryptor.go]: goimports leaves this alone unless I put annotations on the Data and Signature fields as well. I'll just move this up under the Labels field to match conventions elsewhere.

File: cmd/ro/main.go [hunk 7]
```diff
@@ -207,6 +228,85 @@ func runDecrypt() {
...
+	encBytes, err := base64.StdEncoding.DecodeString(string(inBytes))
+	if err != nil {
+		log.Println("failed to base64 decode the data, proceeding with raw data")
+		encBytes = inBytes
+	}
+
+	inBytes, err = ioutil.ReadFile(pubKeyPath)
+	processError(err)
+
+	pubKey, _, _, _, err := ssh.ParseAuthorizedKey(inBytes)
```
Comment (kisom) [inline cmd/ro/main.go]: Why are you skipping the error check here?

File: cmd/ro/main.go [hunk 8]
```diff
@@ -10,16 +10,24 @@ import (
 	"os"
 	"strings"
 
+	"net"
```
Comment (kisom) [inline cmd/ro/main.go]: The imports here look off. If you delete the import block and run goimports on it, that should fix it.

**Comment by andrewbuss** [comment]
Thanks for the comments; I think I've addressed them with the latest set of commits.

File: cmd/ro/gopass/gopass.go [hunk 9]
```diff
@@ -91,11 +91,11 @@ func echoOff(fd []uintptr) (int, error) {
 
 // echoOn turns back on the terminal echo.
 func echoOn(fd []uintptr) {
-  // Turn on the terminal echo.
-  pid, e := syscall.ForkExec(sttyArg0, sttyArgvEOn, &syscall.ProcAttr{Dir: exec_cwdir, Files: fd})
-  if e == nil {
-    syscall.Wait4(pid, &ws, 0, nil)
-  }
+	// Turn on the terminal echo.
+	pid, e := syscall.ForkExec(sttyArg0, sttyArgvEOn, &syscall.ProcAttr{Dir: exec_cwdir, Files: fd})
```
Comment (j-delaney) [inline cmd/ro/gopass/gopass.go]: For consistency with the rest of the project it might make sense to change `e` to `err`

File: cmd/ro/main.go [hunk 10]
```diff
@@ -56,17 +66,25 @@ func registerFlags() {
...
-		fmt.Print("Username:")
-		fmt.Scan(&user)
-		var err error
-		pswd, err = gopass.GetPass("Password:")
-		processError(err)
+	if user == "" {
+		user = os.Getenv(userEnv)
+		if user == "" {
+			fmt.Print("Username:")
+			fmt.Scan(&user)
```
Comment (j-delaney) [inline cmd/ro/main.go]: What would happen if the user just left this blank and hit enter?

File: cmd/ro/gopass/gopass.go (hunk 9 above)
Comment (andrewbuss) [inline cmd/ro/gopass/gopass.go]: Ah; this file didn't have any functional changes; it just got caught up in my gofmt command. I'll remove it from the commit in the next push

File: cmd/ro/main.go (hunk 10 above)
Comment (andrewbuss) [inline cmd/ro/main.go]: Scan blocks until non-whitespace is entered; the same behavior as before. Should this be changed?

Comment (j-delaney) [inline cmd/ro/main.go]: I think it's good. I was just curious. Thanks!

File: index.html [hunk 11]
```diff
@@ -492,6 +498,11 @@ <h1 class="page-header">Red October Management</h1>
 					data.Labels[i] = data.Labels[i].trim();
 					if (data.Labels[i] == "") { data.Labels.splice(i, 1); }
 				}
+				data.Usages = data.Usages.split(',');
+				for(var i=0, l=data.Usages.length; i<l; i++){
+					data.Usages[i] = data.Usages[i].trim();
+					if (data.Usages[i] == "") { data.Usages.splice(i, 1); }
```
Comment (j-delaney) [inline index.html]: `data.Usages[i] == ""` should be `data.Usages[i] === ""`. JavaScript is weird with equality.

    0 == "" // Returns true
    [] == "" // Returns true

Comment (codyli520) [inline index.html]: "==="  checks for value equality and type equality

Comment (j-delaney) [inline index.html]: Also, splicing changes the length of the list. For example if usages was `["", "hello", "", "", "world", ""]` it would end up as `["hello", "", "world"]` since not every element would be checked.

File: redoctober.go [hunk 12]
```diff
@@ -765,6 +772,11 @@ var indexHtml = []byte(`<!DOCTYPE html>
 					data.Labels[i] = data.Labels[i].trim();
 					if (data.Labels[i] == "") { data.Labels.splice(i, 1); }
 				}
+				data.Usages = data.Usages.split(',');
+				for(var i=0, l=data.Usages.length; i<l; i++){
+					data.Usages[i] = data.Usages[i].trim();
+					if (data.Usages[i] == "") { data.Usages.splice(i, 1); }
```
Comment (j-delaney) [inline redoctober.go]: See my previous comments about the JS

File: index.html (hunk 11 above)
Comment (andrewbuss) [inline index.html]: Yeah, agreed. I'm not sure this is enough of a problem to warrant fixing with this PR though. We know that the elements of data.Usages will be strings, although I wouldn't put it past javascript to have .. exciting .. special cases of .split().

File: redoctober.go (hunk 12 above)
Comment (andrewbuss) [inline redoctober.go]: BTW, is there a workflow for updating this automatically from index.html? I read more about the rice tool, but it seems that's not used here.

Comment (j-delaney) [inline redoctober.go]: I think someone is working on one. Not sure who though

File: index.html (hunk 11 above)
Comment (andrewbuss) [inline index.html]: Re the splicing; I'm not sure if that's a problem. Can someone with more JS experience advise whether and how this should be fixed? I simply copied from the Labels splitting above. Maybe there should be a separate PR to address all the frontend nits

Comment (j-delaney) [inline index.html]: I've run into this same issue before. You can address it by changing it to:

    for(var i=0; i < data.Usages.length; i++) {
        data.Usages[i] = data.Usages[i].trim();
        if (data.Usages[i] === "") {
            data.Usages.splice(i, 1);
            i--;
        }
    }

I agree that it should be a separate PR. I'll create an issue for it and work on that

Comment (j-delaney) [inline index.html]: Created PR https://github.com/cloudflare/redoctober/pull/116 for it

File: redoctober.go (hunk 12 above)
Comment (grittygrease) [inline redoctober.go]: Nobody is updating Red October to use rice at the moment. We switched cfssl to rice but haven't done it here.

**Comment by andrewbuss** [comment]
Ah, good catch about computeHMAC; pushing a fix in a moment. I didn't realize there were two layers of signatures. Why is the inner signature required in addition to lock/unlock?

**Comment by andrewbuss** [comment]
Does mac.write insert some sort of separator between calls? I'd be worried that labels:["blue","red"] would be modifiable to labels:["bluered"]

**Comment by andrewbuss** [comment]
I've renamed DecryptSign to SSHSignWith - hopefully that will be less confusing and more consistent. The Usages list may have values corresponding to API endpoints - for now only ssh-sign-with and decrypt.

File: README.md [hunk 13]
```diff
@@ -260,3 +260,24 @@ example uses JavaScript's `btoa` and `atob` functions for string
...
+
+Encrypt with the "ssh-sign-with" usage only:
+
+    $ ro -minimum 2 -owners alice,bob -usages ssh-sign-with \
+         -server ro.local -in id_rsa -out id_rsa.encrypted encrypt
+
+Use the remote server to authenticate to an SSH server
+
+    $ ro -server ro.local -in id_rsa.encrypted -pubkey id_rsa.pub ssh root@gibson
+
```
Comment (Bren2010) [inline README.md]: Trailing whitespace.

File: core/core.go [hunk 14]
```diff
@@ -535,11 +552,23 @@ func Decrypt(jsonIn []byte) ([]byte, error) {
...
 	if err != nil {
 		return jsonStatusError(err)
 	}
 
+	// a file must be marked as usable for "decrypt" before we will decrypt
+	// it for the user. If no usages are provided, we permit decryption
+	foundDecrypt := false
+	for _, usage := range usages {
+		if usage == "decrypt" {
+			foundDecrypt = true
```
Comment (Bren2010) [inline core/core.go]: Trivial optimization: You can break here.  You can also wrap this in a if len(usages) != 0 instead of check below.

usages is sorted by Encrypt, right?  What about sort.SearchStrings?

Comment (andrewbuss) [inline core/core.go]: Ah good point; no reason not to break once we find "decrypt". Fixed in the next commit; and for the "ssh-sign-with" check as well. 

But the interface for SearchStrings is not convenient here; we'd need to first check whether the returned index is equal to the length of the list, then check if the string at that index is equal to "decrypt". Usages will be a very small list bounded in size by the number of API endpoints so it's probably not worth optimizing this.

File: core/core.go [hunk 15]
```diff
@@ -554,6 +583,66 @@ func Decrypt(jsonIn []byte) ([]byte, error) {
...
+	data, names, usages, secure, err := crypt.Decrypt(s.Data, s.Name)
+	if err != nil {
+		return jsonStatusError(err)
+	}
+
+	// a file must be marked as usable for "sign" before we will try to sign with it
+	foundSign := false
+	for _, usage := range usages {
+		if usage == "ssh-sign-with" {
+			foundSign = true
```
Comment (Bren2010) [inline core/core.go]: Same thought as above.

**Comment by Bren2010** [comment]
For those interested, this is what the SSH agent sends to the RO server:
screen shot 2015-12-08 at 1 27 50 pm
[RFC 4252]

All of this is public information, with perhaps the exception of the user name.

LGTM
@andrewbuss, :+1:

File: README.md [hunk 16]
```diff
@@ -260,3 +260,23 @@ example uses JavaScript's `btoa` and `atob` functions for string
 conversion. For dealing with files directly, using the
 [HTML5 File API](https://developer.mozilla.org/en-US/docs/Web/API/FileReader.readAsDataURL)
 would be a good option.
+
+## SSH key encryption
```
Comment (j-delaney) [inline README.md]: Good job with this description. I like how concise it is while still showing the user what to do

File: client/client.go [hunk 17]
```diff
@@ -251,6 +251,22 @@ func (c *RemoteServer) Decrypt(req core.DecryptRequest) (*core.ResponseData, err
...
+		return nil, err
+	}
+
+	respBytes, err := c.doAction("ssh-sign-with", reqBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return unmarshalResponseData(respBytes)
+
```
Comment (j-delaney) [inline client/client.go]: [nit] Unecessary newline

File: cmd/ro/main.go [hunk 18]
```diff
@@ -207,6 +227,85 @@ func runDecrypt() {
...
+	p, err = os.StartProcess(sshPath, flag.Args(),
+		&os.ProcAttr{
+			Files: []*os.File{os.Stdin, os.Stdout, os.Stderr},
+		},
+	)
+	if err != nil {
+		log.Fatal("error starting ssh ", err)
+	}
+
+	_, err = p.Wait()
```
Comment (j-delaney) [inline cmd/ro/main.go]: Think there would ever be a case where it would useful to report on exit code not being 0?

File: cmd/ro/roagent/roagent.go [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+		Data:     signer.encryptedKey,
+		TBSData:  msg,
+	}
+
+	resp, err := signer.server.SSHSignWith(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Status != "ok" {
+		log.Fatal("response status error:", resp.Status)
```
Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: `log.Fatal` logs and then calls `os.Exit(1)` s the `return` statement following this line will never be executed

File: cmd/ro/roagent/roagent.go [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+		return nil, errors.New("response status error")
+	}
+
+	var respMsg core.SSHSignatureWithDelegates
+	err = json.Unmarshal(resp.Response, &respMsg)
+	if err != nil {
+		return nil, err
+	}
+	return &respMsg.Signature, nil
```
Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: You can get rid of the `if err != nil` block directly above this and just do `return &respMsg.Signature, err`

File: cmd/ro/roagent/roagent.go [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+// Add has no effect for the ROAgent
+func (r *ROAgent) Add(key agent.AddedKey) error {
+	return nil
+}
+
+// Sign returns a signature for the data.
+func (r *ROAgent) Sign(key ssh.PublicKey, data []byte) (*ssh.Signature, error) {
+	wanted := key.Marshal()
+	if bytes.Equal(r.signer.PublicKey().Marshal(), wanted) {
+		return r.signer.Sign(rand.Reader, data)
```
Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: Should this be `return r.signer.Sign(rand.Reader, data), nil` since this function is expecting two values to be returned?

File: cryptor/cryptor.go [hunk 22]
```diff
@@ -195,6 +200,10 @@ func (encrypted *EncryptedData) computeHmac(key []byte) []byte {
 		mac.Write([]byte(encrypted.Labels[index]))
 	}
 
+	for index := range encrypted.Usages {
```
Comment (j-delaney) [inline cryptor/cryptor.go]: You could also do:

    for _, value := range encrypted.Usages {
        mac.Write([]byte(value))
    }

File: cmd/ro/roagent/roagent.go (hunk 21 above)
Comment (Bren2010) [inline cmd/ro/roagent/roagent.go]: Sign returns two arguments.

Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: Oops! Didn't realize that

File: cmd/ro/main.go (hunk 18 above)
Comment (andrewbuss) [inline cmd/ro/main.go]: I'm not sure - do others have thoughts either way? My only thought is that if this is to be a drop-in wrapper for SSH, we wouldn't want to add any extra output in the event of an error. If you run `ssh user@host command`, ssh returns the exit code of command. If we wanted to report the error, I'd rather do it by exiting ro with the same exit code as ssh.

File: cmd/ro/roagent/roagent.go (hunk 20 above)
Comment (andrewbuss) [inline cmd/ro/roagent/roagent.go]: I'm hesitant to return a non-nil value if an error occurs - nil is not the same as a pointer to an uninitialized struct.

Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: Good point

File: cmd/ro/roagent/roagent.go (hunk 19 above)
Comment (andrewbuss) [inline cmd/ro/roagent/roagent.go]: This is fixed; though github didn't mark this as a "comment on an outdated diff"

Comment (j-delaney) [inline cmd/ro/roagent/roagent.go]: That's because GitHub does "outdated" by whether the line being commented on has changed, not the file itself.

**Comment by andrewbuss** [comment]
Anything I need to do to make this pullable? No rush; just want to make sure you aren't waiting on any changes from me.

**Comment by grittygrease** [comment]
This is looking good. What are the changes golint is recommending, @andrewbuss?

**Comment by andrewbuss** [comment]
Here's a commit which fixes some of the easiest: https://github.com/andrewbuss/redoctober/commit/c553dd70bbdba0015ba7b25b838163cb24d948ba

It recommends mostly docstrings, but that requires more than a few seconds' thought to fix.

**Comment by grittygrease** [comment]
These all look fine. Feel free to add these to the PR. I think it's close to being merged.

**Comment by andrewbuss** [comment]
It's a little behind master and doesn't merge cleanly with this decrypt_sign branch so I'd rather make it a separate PR once this is merged. It touches a bunch of files and would clutter up the PR diff.

**Comment by grittygrease** [comment]
Fair enough. Feel free to close this and open a new PR.

**Comment by andrewbuss** [comment]
I've rebased on master again; is this good to merge?

File: README.md [hunk 23]
```diff
@@ -341,3 +341,23 @@ example uses JavaScript's `btoa` and `atob` functions for string
...
+    $ ssh-keygen -f id_rsa -N ""
+
+Encrypt with the "ssh-sign-with" usage only:
+
+    $ ro -minimum 2 -owners alice,bob -usages ssh-sign-with \
+         -server ro.local -in id_rsa -out id_rsa.encrypted encrypt
+
+Use the remote server to authenticate to an SSH server
+
+    $ ro -server ro.local -in id_rsa.encrypted -pubkey id_rsa.pub ssh root@gibson
```
Comment (grittygrease) [inline README.md]: You are going to need to add "pubkey" in the commandSet set for ro.

**Comment by grittygrease** [comment]
@andrewbuss The good news is that I tested this and on the happy path, it works. The bad news is that unhappy path could use some polish.

1) When the connection to the server fails, the user experience could be improved. For example, I ran into some issues with setting this up locally because of the fact I didn't have a trusted certificate for localhost. Upon connecting, the ro client returned:
2016/02/01 19:05:04 Request to server failed. Will try again in 5m0s

Luckily, I was able to set up a quick PKI using CFSSL and fix the issue, but not after looking at the red october server logs.

However, some of the more common use cases returned decent enough errors:

> 2016/02/01 19:05:51 agent 13: Need more delegated keys
> 2016/02/01 19:07:18 agent 13: Wrong Password

Although I'm not sure why it says "agent 13," maybe you could help clarify that or make the error more informational.

Also on the more informative cases, the ro client did not go into a connection retry loop. Instead, it tried to connect to the server with my other local SSH keys. I'm not sure that's intended.

2) When the key of a different format, such as ed25519, odd things happen

I tried encrypting a password-encrypted ed25519 SSH key with red october and using it to connect to an SSH server that had password access enabled, but it ended up returning
2016/02/01 18:50:53 failed to parse SSH public keyssh: no key found

and then printed out that the system asked me for a secondary SSH key for the server
`$ Enter passphrase for key "<>" `
and then seemed to fill in some answers for the server's password requests
nick@<server>'s password:
Permission denied, please try again.
nick@<server>'s password:
Permission denied, please try again.
nick@<server>'s password:
Permission denied (publickey,password).

Maybe failing out completely when the key comes back 

a) be more explicit about which key types are supported in the README and docs, and return an appropriate error when an unsupported key is used (rather than failed to parse).
b) Integrate support for passphrase-encrypted SSH keys - we have the gopass module integrated, or at least identify when an user attempts to use a passphrase-encrypted private key and let them know it's not supported.


Things are looking good, just a little more polish on the error paths. Unit tests should make these error cases easier to diagnose and test. Great work.

**Comment by andrewbuss** [comment]
Thanks for the feedback! I probably won't be able to get to this for several weeks unfortunately - heavy course load. Ping me here in late March if there's been no movement.

**Comment by kisom** [comment]
This PR has merge conflicts and is unfinished. Closing it, feel free to re-open later.
